src/audio/module_adapter/module/volume/volume.c

```diff
@@ -293,49 +291,39 @@ static void volume_ramp(struct processing_module *mod)
 
 	/* Current ramp time in Q29.3 milliseconds. Note that max. ramp length
 	 * can be 1.3s at 192 kHz rate and 5.5s at 48 kHz rate without
-	 * exceeding int32_t range. Value 8000 is 1000 for converting to
-	 * milliseconds times 8 (2 ^ 3) for fraction.
+	 * exceeding int32_t range. Inverse of sample rate is 1000/sample_rate
+	 * for milliseconds.
 	 */
-	ramp_time = cd->vol_ramp_elapsed_frames * 8000 / cd->sample_rate;
+	ramp_time = Q_MULTSR_32X32((int64_t)cd->vol_ramp_elapsed_frames,
+				   cd->sample_rate_inv, 0, 31, 3);
 
 	/* Update each volume if it's not at target for active channels */
 	for (i = 0; i < cd->channels; i++) {
 		/* skip if target reached */
-		if (cd->volume[i] == cd->tvolume[i])
+		volume = cd->volume[i];
+		tvolume = cd->tvolume[i];
+		if (volume == tvolume)
 			continue;
 
 		/* Update volume gain with ramp. The ramp gain value is
 		 * calculated from previous gain and ramp time. The slope
 		 * coefficient is calculated in volume_set_chan().
 		 */
-		vol = cd->ramp_func(mod, ramp_time, i);
-		if (cd->volume[i] < cd->tvolume[i]) {
+		new_vol = cd->ramp_func(cd, ramp_time, i);
+		if (volume < tvolume) {
 			/* ramp up, check if ramp completed */
-			if (vol >= cd->tvolume[i] || vol >= cd->vol_max) {
-				cd->ramp_coef[i] = 0;
-				cd->volume[i] = cd->tvolume[i];
-			} else {
+			if (new_vol >= tvolume)
```

Comment:
ramp_func should make sure this can't be happen.

Reply:
Ramp_func does not use tvolume, but rvolume, it may hit exactly tvolume or exceed it slightly due limited precision. I don't see improvement from moving this check to ramp_func. Or is there, how?

Reply:
if slightly exceed, is this also expected? I mean it may not have impact on final volume, 10000/10001, there is no difference for volume output.

I am also fine to keep this.

Reply:
The ramp time is quantized to multiple of 125 us, so the final point can be off a bit. The more, the steeper the ramp is. The check if the ramp is up or down is needed to clamp the gain to not go over or under so I don't see simplification possibility for this.